## Week 5 Homework 

In this homework we'll put what we learned about Spark in practice.

For this homework we will be using the FHVHV 2021-06 data found here. [FHVHV Data](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz )

In [6]:
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz

--2023-03-05 18:12:06--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz
Resolving github.com (github.com)... 140.82.121.4
Connecting to github.com (github.com)|140.82.121.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/4564ad9e-a6da-4923-ad6f-35ff02446a51?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230305%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230305T181206Z&X-Amz-Expires=300&X-Amz-Signature=90a6b90b87dcadf27d6fd292d38ae9212698ac9ed991fa5081232ac5fce9c62b&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dfhvhv_tripdata_2021-06.csv.gz&response-content-type=application%2Foctet-stream [following]
--2023-03-05 18:12:06--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/4564ad9e

### Question 1: 

**Install Spark and PySpark** 

- Install Spark
- Run PySpark
- Create a local spark session
- Execute spark.version.

What's the output?
- 3.3.2
- 2.1.4
- 1.2.3
- 5.4

In [1]:
import pyspark
from pyspark.sql import SparkSession

In [4]:
spark = SparkSession.builder \
    .master("local[*]") \
    .appName('homework') \
    .getOrCreate()

23/03/05 18:09:35 WARN SparkSession: Using an existing Spark session; only runtime SQL configurations will take effect.


In [5]:
spark.version

'3.3.2'

#### Answer: 3.3.2

### Question 2: 

**HVFHW June 2021**

Read it with Spark using the same schema as we did in the lessons.</br> 
We will use this dataset for all the remaining questions.</br>
Repartition it to 12 partitions and save it to parquet.</br>
What is the average size of the Parquet (ending with .parquet extension) Files that were created (in MB)? Select the answer which most closely matches.</br>


- 2MB
- 24MB
- 100MB
- 250MB

In [9]:
from pyspark.sql import types

In [11]:
fhvhv_schema = types.StructType([
    types.StructField("dispatching_base_num", types.StringType(), True),
    types.StructField("pickup_datetime", types.TimestampType(), True),
    types.StructField("dropoff_datetime", types.TimestampType(), True),
    types.StructField("PULocationID", types.IntegerType(), True),
    types.StructField("DOLocationID", types.IntegerType(), True),
    types.StructField("SR_Flag", types.StringType(), True),
    types.StructField("Affiliated_base_number", types.StringType(), True)
])

In [12]:
df = spark.read \
    .option("header", "true") \
    .schema(fhvhv_schema) \
    .csv('fhvhv_tripdata_2021-06.csv.gz')

In [13]:
df.show()

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|Affiliated_base_number|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|              B02764|2021-06-01 00:02:41|2021-06-01 00:07:46|         174|          18|      N|                B02764|
|              B02764|2021-06-01 00:16:16|2021-06-01 00:21:14|          32|         254|      N|                B02764|
|              B02764|2021-06-01 00:27:01|2021-06-01 00:42:11|         240|         127|      N|                B02764|
|              B02764|2021-06-01 00:46:08|2021-06-01 00:53:45|         127|         235|      N|                B02764|
|              B02510|2021-06-01 00:45:42|2021-06-01 01:03:33|         144|         146|      N|                  null|
|              B02510|2021-06-01 00:18:1

In [15]:
df \
    .repartition(12) \
    .write.parquet('fhvhv_tripdata_2021-06')

In [17]:
!ls -lh ./fhvhv_tripdata_2021-06/

total 284M
-rw-r--r-- 1 ArtemiyNaumov ArtemiyNaumov   0 Mar  5 18:20 _SUCCESS
-rw-r--r-- 1 ArtemiyNaumov ArtemiyNaumov 24M Mar  5 18:20 part-00000-4e86b37c-575e-43f1-9a86-9ca91d23fca0-c000.snappy.parquet
-rw-r--r-- 1 ArtemiyNaumov ArtemiyNaumov 24M Mar  5 18:20 part-00001-4e86b37c-575e-43f1-9a86-9ca91d23fca0-c000.snappy.parquet
-rw-r--r-- 1 ArtemiyNaumov ArtemiyNaumov 24M Mar  5 18:20 part-00002-4e86b37c-575e-43f1-9a86-9ca91d23fca0-c000.snappy.parquet
-rw-r--r-- 1 ArtemiyNaumov ArtemiyNaumov 24M Mar  5 18:20 part-00003-4e86b37c-575e-43f1-9a86-9ca91d23fca0-c000.snappy.parquet
-rw-r--r-- 1 ArtemiyNaumov ArtemiyNaumov 24M Mar  5 18:20 part-00004-4e86b37c-575e-43f1-9a86-9ca91d23fca0-c000.snappy.parquet
-rw-r--r-- 1 ArtemiyNaumov ArtemiyNaumov 24M Mar  5 18:20 part-00005-4e86b37c-575e-43f1-9a86-9ca91d23fca0-c000.snappy.parquet
-rw-r--r-- 1 ArtemiyNaumov ArtemiyNaumov 24M Mar  5 18:20 part-00006-4e86b37c-575e-43f1-9a86-9ca91d23fca0-c000.snappy.parquet
-rw-r--r-- 1 ArtemiyNaumov ArtemiyNaumov

#### Answer: 24MB

### Question 3: 

**Count records**  

How many taxi trips were there on June 15?</br></br>
Consider only trips that started on June 15.</br>

- 308,164
- 12,856
- 452,470
- 50,982

In [18]:
df.registerTempTable('fhvhv_trips_data')

/home/ArtemiyNaumov/spark/spark-3.3.2-bin-hadoop3/python/pyspark/sql/dataframe.py:229: FutureWarning: Deprecated in 2.0, use createOrReplaceTempView instead.
  warnings.warn("Deprecated in 2.0, use createOrReplaceTempView instead.", FutureWarning)


In [21]:
query = '''
    SELECT 
        COUNT(1)
    FROM fhvhv_trips_data
    WHERE CAST(pickup_datetime AS DATE) = '2021-06-15'
'''

In [22]:
spark.sql(query).show()

+--------+
|count(1)|
+--------+
|  452470|
+--------+



In [27]:
from datetime import datetime
from pyspark.sql import functions as F

In [31]:
df.filter(F.to_date(F.col('pickup_datetime')) == F.lit('2021-06-15')).count()

452470

#### Answer: 452,470

### Question 4: 

**Longest trip for each day**  

Now calculate the duration for each trip.</br>
How long was the longest trip in Hours?</br>

- 66.87 Hours
- 243.44 Hours
- 7.68 Hours
- 3.32 Hours

In [41]:
query_2 = '''
    SELECT
        CAST(TIMESTAMPDIFF(SECOND, pickup_datetime, dropoff_datetime) AS NUMERIC) / 3600 AS trip_duration_in_hours
    FROM fhvhv_trips_data
    ORDER BY 1 DESC
'''

In [42]:
spark.sql(query_2).show()

+----------------------+
|trip_duration_in_hours|
+----------------------+
|             66.878889|
|             25.549722|
|             19.980833|
|             18.197222|
|             16.466944|
|             14.268889|
|             13.909722|
|             11.670000|
|             11.365833|
|             10.984444|
|             10.267500|
|              9.966389|
|              9.966389|
|              9.637778|
|              9.624444|
|              9.480278|
|              9.471667|
|              9.402222|
|              9.393611|
|              9.376944|
+----------------------+
only showing top 20 rows



#### Answer: 66.87 Hours

### Question 5: 

**User Interface**

 Spark’s User Interface which shows application's dashboard runs on which local port?</br>

- 80
- 443
- 4040
- 8080

#### Answer: 4040

### Question 6: 

**Most frequent pickup location zone**

Load the zone lookup data into a temp view in Spark</br>
[Zone Data](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv)</br>

Using the zone lookup data and the fhvhv June 2021 data, what is the name of the most frequent pickup location zone?</br>

- East Chelsea
- Astoria
- Union Sq
- Crown Heights North

In [43]:
df_zones = spark.read \
    .option("header", "true") \
    .csv('taxi_zone_lookup.csv')

In [45]:
df_zones.printSchema()

root
 |-- LocationID: string (nullable = true)
 |-- Borough: string (nullable = true)
 |-- Zone: string (nullable = true)
 |-- service_zone: string (nullable = true)



In [46]:
df_zones.registerTempTable('zones')

/home/ArtemiyNaumov/spark/spark-3.3.2-bin-hadoop3/python/pyspark/sql/dataframe.py:229: FutureWarning: Deprecated in 2.0, use createOrReplaceTempView instead.
  warnings.warn("Deprecated in 2.0, use createOrReplaceTempView instead.", FutureWarning)


In [51]:
query_3 = '''
    SELECT 
        pickup_zone.Zone,
        count(1) as cnt_
    FROM fhvhv_trips_data
    JOIN zones AS pickup_zone ON
        fhvhv_trips_data.PULocationID = pickup_zone.LocationID
    GROUP BY 1 ORDER BY 2 DESC
'''

In [52]:
spark.sql(query_3).show()

+--------------------+------+
|                Zone|  cnt_|
+--------------------+------+
| Crown Heights North|231279|
|        East Village|221244|
|         JFK Airport|188867|
|      Bushwick South|187929|
|       East New York|186780|
|TriBeCa/Civic Center|164344|
|   LaGuardia Airport|161596|
|            Union Sq|158937|
|        West Village|154698|
|             Astoria|152493|
|     Lower East Side|151020|
|        East Chelsea|147673|
|Central Harlem North|146402|
|Williamsburg (Nor...|143683|
|          Park Slope|143594|
|  Stuyvesant Heights|141427|
|        Clinton East|139611|
|West Chelsea/Huds...|139431|
|             Bedford|138428|
|         Murray Hill|137879|
+--------------------+------+
only showing top 20 rows



#### Answer: Crown Heights North